### PR TITLE
Limit typeparams in Generic Fake creation methods to reference types

### LINKED ISSUE #1465: Limit typeparams in Generic Fake creation methods to reference types
I don't think this has been discussed before, but maybe I'm forgetting.

The `A.Fake` method could be defined as 

```c#
public static T Fake<T>() where T : class
```

(As could all other generic Fake creation methods.)

That way users would be prevented _by the compiler_ from trying to fake a struct or similar.

The helpful error message we output when they _do_ try doesn't really provide a lot more information than the typeparam constraint would:

```
Failed to create fake of type FakeItEasy.Specs.CreationSpecsBase+Struct:
  The type of proxy must be an interface or a class but it was FakeItEasy.Specs.CreationSpecsBase+Struct.
```

so I'm not sure we'd be losing much, if anything.

### PR DESCRIPTION
**Pull request by blairconrad**
Fixes #1465.

### CODE REVIEW

File: src/FakeItEasy/Core/DelegateRaiser.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ namespace FakeItEasy.Core
     /// in order to raise that event.
     /// </summary>
     /// <typeparam name="TEventHandler">The type of the event handler. Should be a <see cref="Delegate"/></typeparam>
-    internal class DelegateRaiser<TEventHandler> : IEventRaiserArgumentProvider
+    internal class DelegateRaiser<TEventHandler> : IEventRaiserArgumentProvider where TEventHandler : class
```
[the remark this answers is not in the document]
Comment (blairconrad) [reply src/FakeItEasy/Core/DelegateRaiser.cs]: Oh, I'd tried originally and got an error. Will try again.

[the remark this answers is not in the document]
Comment (thomaslevesque) [reply src/FakeItEasy/Core/DelegateRaiser.cs]: Strange, I tried and it worked...

**Comment by blairconrad** [comment]
Squashed. I left the unsquashed version in https://github.com/blairconrad/FakeItEasy/tree/dont-fake-a-value-unsquashed, in case you want to compare.

**Comment by thomaslevesque** [comment]
Thanks, @blairconrad!

**Comment by blairconrad** [comment]
Thank you!

**Pull request merged**